# Amazon Dataset

In [2]:
pip install mysql-connector-python

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [3]:
#Import necessay libraries
import numpy as np 
import pandas as pd
import mysql.connector

In [4]:
# MySQL database connection configuration
db_config = {
    'host': 'localhost',
    'user': 'root',
    'password': 'kenyan23'
}

In [5]:
# Establish connection to MySQL database
conn = mysql.connector.connect(**db_config)

In [6]:
# Function to execute MySQL queries
def execute_query(query):
    cursor = conn.cursor()
    cursor.execute(query)
    conn.commit()
    cursor.close()

## Data Preprocessing

### Loading the Dataset

In [7]:
#loading dataset
df = pd.read_csv('amazon.csv')

Exploring Dataset

In [8]:
#Get information of the no. of rows and columns
df.shape

(1465, 16)

In [9]:
#titles of the 12 columns
print (df.columns.values)

['product_id' 'product_name' 'category' 'discounted_price' 'actual_price'
 'discount_percentage' 'rating' 'rating_count' 'about_product' 'user_id'
 'user_name' 'review_id' 'review_title' 'review_content' 'img_link'
 'product_link']


In [10]:
df.head(5)

product_id                                       product_name  \
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   

                                            category discounted_price  \
0  Computers&Accessories|Accessories&Peripherals|...             ₹399   
1  Computers&Accessories|Accessories&Peripherals|...             ₹199   
2  Computers&Accessories|Accessories&Peripherals|...             ₹199   
3  Computers&Accessories|Accessories&Peripherals|...             ₹329   
4  Computers&Accessories|Accessories&Peripherals|...             ₹154   

  actual_price discount_percentage rating rating_count  \
0       ₹1,099                 64%    4.2       24,269   
1         ₹349                 43%    4.0       43,994   
2       ₹1,899                 90%    3.9        7,928   
3         ₹699                 53%    4.2       94,363   
4         ₹399                 61%    4.2       16,905   

                                       about_product  \
0  High Compatibility : Compatible With iPhone 12...   
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   

                                             user_id  \
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...   
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   

                                           user_name  \
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...   
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   

                                           review_id  \
0  R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...   
1  RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...   
2  R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...   
3  R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...   
4  R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...   

                                        review_title  \
0  Satisfied,Charging is really fast,Value for mo...   
1  A Good Braided Cable for Your Type C Device,Go...   
2  Good speed for earlier versions,Good Product,W...   
3  Good product,Good one,Nice,Really nice product...   
4  As good as original,Decent,Good one for second...   

                                      review_content  \
0  Looks durable Charging is fine tooNo complains...   
1  I ordered this cable to connect my phone to An...   
2  Not quite durable and sturdy,https://m.media-a...   
3  Good product,long wire,Charges good,Nice,I bou...   
4  Bought this instead of original apple, does th...   

                                            img_link  \
0  https://m.media-amazon.com/images/W/WEBP_40237...   
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

In [11]:
df.tail(10)

product_id                                       product_name  \
1455  B00GHL8VP2  USHA 1212 PTC with Adjustable Thermostat Fan H...   
1456  B0B9JZW1SQ  4 in 1 Handheld Electric Vegetable Cutter Set,...   
1457  B00TI8E7BI  Philips HD9306/06 1.5-Litre Electric Kettle (M...   
1458  B07J9KXQCC  Libra Room Heater for Home, Room Heaters Home ...   
1459  B0B3JSWG81  NGI Store 2 Pieces Pet Hair Removers for Your ...   
1460  B08L7J3T31  Noir Aqua - 5pcs PP Spun Filter + 1 Spanner | ...   
1461  B01M6453MB  Prestige Delight PRWO Electric Rice Cooker (1 ...   
1462  B009P2LIL4  Bajaj Majesty RX10 2000 Watts Heat Convector R...   
1463  B00J5DYCCA  Havells Ventil Air DSP 230mm Exhaust Fan (Pist...   
1464  B01486F4G6  Borosil Jumbo 1000-Watt Grill Sandwich Maker (...   

                                               category discounted_price  \
1455  Home&Kitchen|Heating,Cooling&AirQuality|RoomHe...        ₹3,487.77   
1456  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...             ₹498   
1457  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...           ₹2,695   
1458  Home&Kitchen|Heating,Cooling&AirQuality|RoomHe...             ₹949   
1459  Home&Kitchen|Kitchen&HomeAppliances|Vacuum,Cle...             ₹199   
1460  Home&Kitchen|Kitchen&HomeAppliances|WaterPurif...             ₹379   
1461  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...           ₹2,280   
1462  Home&Kitchen|Heating,Cooling&AirQuality|RoomHe...           ₹2,219   
1463  Home&Kitchen|Heating,Cooling&AirQuality|Fans|E...           ₹1,399   
1464  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...           ₹2,863   

     actual_price discount_percentage rating rating_count  \
1455       ₹4,990                 30%    4.1        1,127   
1456       ₹1,200                 59%    3.2          113   
1457       ₹2,695                  0%    4.4        2,518   
1458       ₹2,299                 59%    3.6          550   
1459         ₹999                 80%    3.1            2   
1460         ₹919                 59%      4        1,090   
1461       ₹3,045                 25%    4.1        4,118   
1462       ₹3,080                 28%    3.6          468   
1463       ₹1,890                 26%      4        8,031   
1464       ₹3,690                 22%    4.3        6,987   

                                          about_product  \
1455  Meant for Spot Heating|Room Size: Upto 15 sq f...   
1456  { 4 in 1 multi-function Electric Vegetable Cut...   
1457  Food grade stainless steel, Frequency: 50 Hz. ...   
1458  Libra room heater for home has cool, warm or h...   
1459  Versatile, Protects Your Clothes: MoveCatcher ...   
1460  SUPREME QUALITY 90 GRAM 3 LAYER THIK PP SPUN F...   
1461                       230 Volts, 400 watts, 1 Year   
1462  International design and styling|Two heat sett...   
1463  Fan sweep area: 230 MM ; Noise level: (40 - 45...   
1464  Brand-Borosil, Specification – 23V ~ 5Hz;1 W...   

                                                user_id  \
1455  AECFYIUCHSZXDLACTYPEUSM5DIKA,AF546S43YKYCJWRD7...   
1456  AFCTMQKPVJI6Y2JPIGDKRKIAV43A,AF6XUHN32GSFA7LFG...   
1457  AHYXOMUJUKZHBWHP43ZAB265EDGA,AG4C27NTZZ7HTG6W3...   
1458  AE5FZ5B3EEES45Q26PNUBTJ5DRYA,AGL2B432J2VDHOSUW...   
1459                       AFPKVN5KLHB4MHSYDS25Q5MIWWKQ   
1460  AHITFY6AHALOFOHOZEOC6XBP4FEA,AFRABBODZJZQB6Z4U...   
1461  AFG5FM3NEMOL6BNFRV2NK5FNJCHQ,AGEINTRN6Z563RMLH...   
1462  AGVPWCMAHYQWJOQKMUJN4DW3KM5Q,AF4Q3E66MY4SR7YQZ...   
1463  AF2JQCLSCY3QJATWUNNHUSVUPNQQ,AFDMLUXC5LS5RXDJS...   
1464  AFGW5PT3R6ZAVQR4Y5MWVAKBZAYA,AG7QNJ2SCS5VS5VYY...   

                                              user_name  \
1455  Manoj Kumar,Arun Thottakathu Prasannakumar,son...   
1456  Ashish,Kavita J.,Sundar,Tessy S.,saurabh manro...   
1457  AmazonCust,DD_Zen,Amazon Customer,VINISH THOMA...   
1458  gitesh jaiswal,Divyansh,Dr Paras,Amazon Custom...   
1459                                          Kunal Sen   
1460  Prabha ds,Raghuram bk,Real Deal,Amazon Custome...   
14

In [12]:
df.describe()

product_id                                       product_name  \
count         1465                                               1465   
unique        1351                                               1337   
top     B07JW9H4J1  Fire-Boltt Ninja Call Pro Plus 1.83" Smart Wat...   
freq             3                                                  5   

                                                 category discounted_price  \
count                                                1465             1465   
unique                                                211              550   
top     Computers&Accessories|Accessories&Peripherals|...             ₹199   
freq                                                  233               53   

       actual_price discount_percentage rating rating_count  \
count          1465                1465   1465         1463   
unique          449                  92     28         1143   
top            ₹999                 50%    4.1        9,378   
freq            120                  56    244            9   

                                            about_product  \
count                                                1465   
unique                                               1293   
top     [CHARGE & SYNC FUNCTION]- This cable comes wit...   
freq                                                    6   

                                                  user_id  \
count                                                1465   
unique                                               1194   
top     AHIKJUDTVJ4T6DV6IUGFYZ5LXMPA,AE55KTFVNXYFD5FPY...   
freq                                                   10   

                                                user_name  \
count                                                1465   
unique                                               1194   
top     $@|\|TO$|-|,Sethu madhav,Akash Thakur,Burger P...   
freq                                                   10   

                                                review_id  \
count                                                1465   
unique                                               1194   
top     R3F4T5TRYPTMIG,R3DQIEC603E7AY,R1O4Z15FD40PV5,R...   
freq                                                   10   

                                             review_title  \
count                                                1465   
unique                                               1194   
top     Worked on iPhone 7 and didn’t work on XR,Good ...   
freq                                                   10   

                                           review_content  \
count                                                1465   
unique                                               1212   
top     I am not big on camera usage, personally. I wa...   
freq                                                    8   

                                                 img_link  \
count                                                1465   
unique                                               1412   
top     https://m.media-amazon.com/images/I/413sCRKobN...   
freq                                                    3   

                                             product_link  
count                                                1465  
unique                                               1465  
top     https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
freq                                                    1

In [13]:
#checking for missing values 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   product_id           1465 non-null   object
 1   product_name         1465 non-null   object
 2   category             1465 non-null   object
 3   discounted_price     1465 non-null   object
 4   actual_price         1465 non-null   object
 5   discount_percentage  1465 non-null   object
 6   rating               1465 non-null   object
 7   rating_count         1463 non-null   object
 8   about_product        1465 non-null   object
 9   user_id              1465 non-null   object
 10  user_name            1465 non-null   object
 11  review_id            1465 non-null   object
 12  review_title         1465 non-null   object
 13  review_content       1465 non-null   object
 14  img_link             1465 non-null   object
 15  product_link         1465 non-null   object
dtypes: obj

In [18]:
#rating count has less values
# Using median to fill in the existing values 
# fill in missing values with the median
# Remove commas (if numbers are formatted with commas)
df['rating_count'] = df['rating_count'].replace(',', '', regex=True)

# Convert to numeric, forcing errors to NaN
df['rating_count'] = pd.to_numeric(df['rating_count'], errors='coerce')

In [19]:
median_rating_count = df['rating_count'].median()
print("Median rating count:", median_rating_count)

Median rating count: 293.0


In [20]:
#confirming if missing values are handled
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   product_id           1465 non-null   object 
 1   product_name         1465 non-null   object 
 2   category             1465 non-null   object 
 3   discounted_price     1465 non-null   object 
 4   actual_price         1465 non-null   object 
 5   discount_percentage  1465 non-null   object 
 6   rating               1465 non-null   object 
 7   rating_count         1465 non-null   float64
 8   about_product        1465 non-null   object 
 9   user_id              1465 non-null   object 
 10  user_name            1465 non-null   object 
 11  review_id            1465 non-null   object 
 12  review_title         1465 non-null   object 
 13  review_content       1465 non-null   object 
 14  img_link             1465 non-null   object 
 15  product_link         1465 non-null   o

Missing values are handled well and all columns have 1,465 values

### Data Cleaning

In [21]:
# Display data types
print(df[['discounted_price', 'actual_price']].dtypes)

# Display first few rows to check for symbols
print(df[['discounted_price', 'actual_price']].head())


discounted_price    object
actual_price        object
dtype: object
  discounted_price actual_price
0             ₹399       ₹1,099
1             ₹199         ₹349
2             ₹199       ₹1,899
3             ₹329         ₹699
4             ₹154         ₹399


In [22]:
# Remove currency symbols and commas, then convert to numeric
df['discounted_price'] = df['discounted_price'].replace('[^\d.]', '', regex=True)
df['actual_price'] = df['actual_price'].replace('[^\d.]', '', regex=True)

# Convert to numeric type
df['discounted_price'] = pd.to_numeric(df['discounted_price'], errors='coerce')
df['actual_price'] = pd.to_numeric(df['actual_price'], errors='coerce')


In [23]:
# Confirm data types
print(df[['discounted_price', 'actual_price']].dtypes)

# Check for NaN values (if any errors occurred)
print(df[['discounted_price', 'actual_price']].isnull().sum())

# Display first few rows
print(df[['discounted_price', 'actual_price']].head())


discounted_price    float64
actual_price        float64
dtype: object
discounted_price    0
actual_price        0
dtype: int64
   discounted_price  actual_price
0             399.0        1099.0
1             199.0         349.0
2             199.0        1899.0
3             329.0         699.0
4             154.0         399.0


Dicounted and actual price has been converted from object type to float64. In addition, all commas and currencies were removed from the columns. There are no missing values in the two columns

In [ ]:
# Create database if not exists
create_db_query = "CREATE DATABASE IF NOT EXISTS ecommerce_db;"
execute_query(create_db_query)


In [ ]:
# Use the database
use_db_query = "USE ecommerce_db;"
execute_query(use_db_query)

In [ ]:
create_raw_data_table_query = """
CREATE TABLE IF NOT EXISTS products (
    product_id INT AUTO_INCREMENT PRIMARY KEY,
    product_name VARCHAR(255),
    category VARCHAR(100),
    discounted_price DECIMAL(10,2),
    actual_price DECIMAL(10,2),
    discount_percentage FLOAT,
    rating FLOAT,
    rating_count INT
)
"""
execute_query(create_raw_data_table_query)